### DISCUSSION TITLE
How do you unit test a Relm4 app?

### DISCUSSION THREAD

**Discussion by gtamas** [Relm4/Relm4#604] [General]
How to do this?
Should I look into something like this?

https://github.com/gtk-rs/gtk-test

Or is there a better way to test GTK4 / Relm4 code?

I wanna write unit tests for my GTK4 helpers, which call GTK nethods and return GTK types (like subclasses of Widget), and I wanna test my relm4 Components / SimpleComponents as well.

Thanks in advance!

**Comment by AaronErhardt** [comment]
The best approach would be to use the [`#[gtk::test]`](https://gtk-rs.org/gtk4-rs/git/docs/gtk4_macros/attr.test.html) attribute on your tests. Relm4 doesn't offer any testing infrastructure yet, but if you come up with a small example or find a use-case that isn't covered properly, let us know!
